source/simulator/core.cc

```diff
@@ -1839,6 +1839,14 @@ namespace aspect
         // calculate global volume after deforming mesh
         global_volume = GridTools::volume (triangulation, *mapping);
         signals.post_mesh_deformation(*this);
+
+        // We update this after mesh deformation so the most recent surface is kept.
```

Comment:
Just call the `update` function unconditionally here. The geometry model interface or the box geometry model is responsible to make sure to behave correctly for each call of `update`. Your current implementation makes assumptions about the implementation of specific plugins in the core part of ASPECT, which means we would have to change the core if the implementation of the plugin changes, which we want to avoid. 
There is still one small problem with the current implementation, which is that now the geometry model `update` function is called twice, once in `start_timestep` and another time here. This contradicts our documentation of the update function, namely that it is called exactly once per time step at the beginning of a new time step. I have to think about this a bit more, and if there is a way to prevent it.